### ISSUE TITLE
Error connecting to AWS-IoT with deviceOS 1.4.0

### ISSUE THREAD

**Issue by mhazley** [hirotakaster/MQTT-TLS#27]
Hi, sorry about how vague this is, I haven't had a lot of time to get deeper on this, I just wanted to flag it up in case anyone has any insights. For now I have rolled back to deviceOS1.1.1 (Just the last one I was using).

When using deviceOS 1.4.0 - I get this logging out of the library:

```
hand shake start
Verify requested for (Depth 2):
cert. version     : 3
serial number     : 06:6C:9F:CF:99:BF:8C:0A:39:E2:F0:78:8A:43:E6:96:36:5B:CA
issuer name       : C=US, O=Amazon, CN=Amazon Root CA 1
subject name      : C=US, O=Amazon, CN=Amazon Root CA 1
issued  on        : 2015-05-26 00:00:00
expires on        : 2038-01-17 00:00:00
signed using      : RSA with SHA-256
RSA key size      : 2048 bits
basic constraints : CA=true
key usage         : Digital Signature, Key Cert Sign, CRL Sign
This certificate has no flags
Verify requested for (Depth 1):
cert. version     : 3
serial number     : 06:7F:94:57:85:87:E8:AC:77:DE:B2:53:32:5B:BC:99:8B:56:0D
issuer name       : C=US, O=Amazon, CN=Amazon Root CA 1
subject name      : C=US, O=Amazon, OU=Server CA 1B, CN=Amazon
issued  on        : 2015-10-22 00:00:00
expires on        : 2025-10-19 00:00:00
signed using      : RSA with SHA-256
RSA key size      : 2048 bits
basic constraints : CA=true, max_pathlen=0
key usage         : Digital Signature, Key Cert Sign, CRL Sign
This certificate has no flags
Verify requested for (Depth 0):
cert. version     : 3
serial number     : 0C:62:FF:6C:06:6E:AD:10:24:28:8A:6C:F1:9D:D4:DC
issuer name       : C=US, O=Amazon, OU=Server CA 1B, CN=Amazon
subject name      : CN=*.iot.eu-west-1.amazonaws.com
issued  on        : 2019-07-02 00:00:00
expires on        : 2020-06-28 12:00:00
signed using      : RSA with SHA-256
RSA key size      : 2048 bits
basic constraints : CA=false
subject alt name  : iot.eu-west-1.amazonaws.com, *.iot.eu-west-1.amazonaws.com
key usage         : Digital Signature, Key Encipherment
ext key usage     : TLS Web Server Authentication, TLS Web Client Authentication
This certificate has no flags
handshake done, ret = -17040
mqtt disconnected
netWrite!!
tls close
```
Sporadically, the `-17040` comes out as `-17168` but not very often. 

I can't seem to find a matching error code in `ssl.h` but when I google around, a lot of people think it has something to do with RAM. 

I know this lib is not an AWS IoT lib but I thought I'd post in case anyone else has the same issue. 

I will hopefully get some time to look a bit deeper into this soon - sorry for the light touch!

**Comment by hirotakaster** [comment]
Hi @mhazley ,

What you use environment? 
My Argon deviceOS 1.4 connect to AWS IoT Core with MQTT-TLS.

17040(0x4290) means  MBEDTLS_ERR_RSA_KEY_CHECK_FAILED and 0x90 error.
The cause of this error are
1. certificate setting
  check your rootcat/csr/private key 
2. device memory
  use low memory version config(config-mini.h) or modify the config.h settings.

thanks.

**Comment by mhazley** [comment]
Hi @hirotakaster!

This was a Boron running deviceOS 1.4.0.

I had modified the `config.h` to remove: 

```
MBEDTLS_KEY_EXCHANGE_ECDHE_ECDSA_ENABLED,MBEDTLS_ASN1_WRITE_C,MBEDTLS_ECDH_C,MBEDTLS_ECDSA_C,MBEDTLS_ECP_C,MBEDTLS_SHA512_C
```

The certs and keys work fine with deviceOS v1.1.1 - so they are fine. 

Perhaps, as this is a Boron, it is using more memory?

I will test more next week when I have some more free time.

**Comment by hirotakaster** [comment]
sorry I don't have a Boron because of my country "Radio Act", I hope that your report.

> Perhaps, as this is a Boron, it is using more memory?

That may be true.

And FYI, here is sample MQTT-TLS application build size on my VScode. Target is my Argon deviceOS 1.4.0.

Default config.h
![config](https://user-images.githubusercontent.com/203933/65736980-ef7db080-e117-11e9-8d36-a98f5ef733e9.png)

config-mini.h
![mini-config](https://user-images.githubusercontent.com/203933/65737009-0a502500-e118-11e9-8585-c4a74b41b1e5.png)

Both sample(config.h/config-mini.h) can communicate with AWS IoT.
config-mini.h can reduce the size&memory but TLS cipher is TLS_RSA_WITH_AES_256_GCM_SHA256 only.

You could copy the config-mini.h to config.h on your Boron app&build, maybe it will work out.

**Comment by mhazley** [comment]
I have done some more work on this. 

I have been using deviceOS 1.1.1 and I have switched to the config-mini.h.

I noticed that as I added some more functionality to my code-base or increased some buffer sizes, I would sometimes cause the library to timeout with the above error.

e.g. it would always time out when this was my program output:

```
Creating /Users/mhazley/src/cat-609/target/1.1.1/boron/module_user_memory.ld ...
   text    data     bss     dec     hex filename
 107956     228   15160  123344   1e1d0
```

But now seems to work after I did some refinement and made a few better design decisions:

```
Creating /Users/mhazley/src/cat-609/target/1.1.1/boron/module_user_memory.ld ...
   text    data     bss     dec     hex
 104788     224   14904  119916   1d46c
```

So... I am not even sure what my issue is/was but it works now so I am going to close.

Thanks for your time.
